Add Support for City of Greater Bendigo VIC, Australia

iSnackTwoPointOh opened the pull request:
# Pull Request: Add Support for City of Greater Bendigo VIC, Australia
## Description
Added support for the City of Greater Bendigo VIC, Australia. 
No ICS file but they use an app called pozi for geographic overlays including the bin schedule so was able to work with that.

## Key Features
Collection types:
- general waste ✅ 
- recycling ✅ 
- green waste ✅ 
- Note: not including hard waste and commercial waste since the data had no entries for either of these (pretty sure Bendigo has no hard waste collection anymore either)

Accepts input coords and checks with the downloaded geometry which schedule they belong to. Orignally used Shapely to process the geomtry but removed that based on feedback and now using a custom ray tracing algo.

## Testing
Test Script: Verified functionality using test_sources.py:
```python
python3 test_sources.py -s bendigo_vic_gov_au      
Testing source bendigo_vic_gov_au ...
  found 3 entries for Bunnings Epsom
  found 3 entries for Bunnings Bendigo
  found 3 entries for Alfa Kitchen Bendigo
```

closes #2258

sagwefwefga commented:
Nice work! assuming we have to wait for them to add/approve this and then update the integration?

iSnackTwoPointOh commented:
@sagwefwefga yeah pretty much. Hopefully there’s not too much of an issue with the added dependency. Otherwise I’ll have to hand roll something

5ila5 commented:
We allready tried adding shaply as dependency in the past and had some trouble with this and decided to remove it again (see  #1705 and #1709)

I'm not the biggest fan of hard-coding the geocode.earth API Key of a commercial third party API

I cannot see the map when visiting the page (probably GeoBlocking issue) So I cannot really look into this.

iSnackTwoPointOh commented:
@5ila5 i've removed the need for the geocode.earth api in favour of just starting from lat/lon. I've also added a ray tracing algo for the geomtry instead of using shapely. Might not be as robust but for a single source its probably good enough

5ila5 commented on custom_components/waste_collection_schedule/waste_collection_schedule/source/bendigo_vic_gov_au.py:
```diff
@@ -0,0 +1,253 @@
+                    f"Longitude must be between {BENDIGO_BOUNDS['min_lon']} and {BENDIGO_BOUNDS['max_lon']}"
+                )
+        except (ValueError, TypeError) as e:
+            raise SourceArgumentNotFound(
+                "coordinates",
+                f"({latitude}, {longitude})",
+                f"Invalid coordinate format. Please provide numeric values. Error: {str(e)}"
+            )
```
This does not work like this just raise a normal exception here

As Ha config will try to match the exception to an input filed to show the error message next to this, but there is no field coordinates 

```suggestion
            raise Exception(
                f"Invalid coordinate format. Please provide numeric values. Error: {str(e)}"
            )
```

5ila5 commented on custom_components/waste_collection_schedule/waste_collection_schedule/source/bendigo_vic_gov_au.py:
```diff
@@ -0,0 +1,253 @@
+        if len(found_zones) > 1:
+            zone_names = [zone["properties"]["name"] for zone in found_zones]
+            _LOGGER.debug("Point found in multiple zones: %s", zone_names)
+            raise SourceArgumentNotFound(
+                "coordinates",
+                f"({self._latitude}, {self._longitude})",
+                f"Coordinates are on a boundary between multiple zones: {', '.join(zone_names)}. "
+                "To resolve this, please try adjusting your coordinates slightly (by 0.0001 degrees or less) "
+                "in any direction. You can verify your zone at "
+                "https://www.bendigo.vic.gov.au/residents/general-waste-recycling-and-organics/bin-night"
+            )
```
same

```suggestion
            raise Exception(
                f"Coordinates ({self._latitude}, {self._longitude}) are on a boundary between multiple zones: {', '.join(zone_names)}. "
                "To resolve this, please try adjusting your coordinates slightly (by 0.0001 degrees or less) "
                "in any direction. You can verify your zone at "
                "https://www.bendigo.vic.gov.au/residents/general-waste-recycling-and-organics/bin-night"
            )
```

5ila5 reviewed (changes requested):
I'm not that big of a fan of all these helper functions inside the fetch function. I think it would be easier to read if you'd put them in the class (maybe marked as @staticmethod) or just above the 

This is ready to merge when this and the above-mentioned Exceptions are changed

iSnackTwoPointOh commented:
> I'm not that big of a fan of all these helper functions inside the fetch function. I think it would be easier to read if you'd put them in the class (maybe marked as @staticmethod) or just above the
> 
> This is ready to merge when this and the above-mentioned Exceptions are changed

agreed with the helpers. I've made them private statics on the class as suggested. Thanks for the feedback :)

5ila5 commented:
Thanks for your contribution

iSnackTwoPointOh commented:
> Unfortunately I believe it’s now broken

Looks like there was an API change recently. I've created #5177  to track. Hopefully ready next couple days.
